### ISSUE TITLE
Ruby Resolv::DNS failing for long DNS records served by CoreDNS

### ISSUE THREAD

**Issue by mattlqx** [coredns/coredns#2318]
I'm experiencing odd behavior with `Resolv::DNS.getaddress` resolving against CoreDNS under certain circumstances. I have a zone served by the route53 plugin and unbound acting as a recursive resolver.

```
example.com {
  route53 example.com.:ZZZZZZ3A0GBKDI example.com.:ZZZZZZ3ABBBBB {
    upstream 127.0.0.1:253  # unbound for recursion
  }
  prometheus
  cache 300
  errors stdout
  log stdout
}
```

Within this zone, I have some CNAMEs that resolve to Amazon ALBs. It appears once the response is beyond a certain size, Ruby's `Resolv::DNS` class fails to recognize the response and retries the request until timeout. I can make the same query to a non-CoreDNS resolver and the call will return with no problem. The server setup is relatively simple: CoreDNS is running on "bare" VMs within a Docker container with "host" networking. 

#### Non-CoreDNS resolver (dnsmasq running off-host)
```
irb(main):003:0> ::Resolv::DNS.new(nameserver: '192.168.2.2').getaddress('service-a.aws.example.com')
=> #<Resolv::IPv4 192.168.44.44>
```

#### CoreDNS resolver running on host
```
irb(main):004:0> ::Resolv::DNS.new(nameserver: '127.0.0.1').getaddress('service-a.aws.example.com')
Resolv::ResolvError: DNS result has no information for service-a.aws.example.com
	from /opt/chef/embedded/lib/ruby/2.4.0/resolv.rb:384:in `getaddress'
	from (irb):4
	from /opt/chef/embedded/bin/irb:11:in `<main>'
```
This returns after a 75 second timeout.

This particular ALB for `service-a` fronts 3 availability zones and thus returns 3 records. `dig` works perfectly fine for this query against both resolvers.

#### Non-CoreDNS
```
; <<>> DiG 9.11.3-1ubuntu1.2-Ubuntu <<>> service-a.aws.example.com @192.168.2.2
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 40578
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;service-a.aws.example.com. IN A

;; ANSWER SECTION:
service-a.aws.example.com. 600 IN	CNAME service-a-1285624372.us-west-2.elb.amazonaws.com.
service-a-1285624372.us-west-2.elb.amazonaws.com. 600 IN A 192.168.44.44
service-a-1285624372.us-west-2.elb.amazonaws.com. 600 IN A 192.168.55.55
service-a-1285624372.us-west-2.elb.amazonaws.com. 600 IN A 192.168.66.66

;; Query time: 0 msec
;; SERVER: 192.168.2.2#53(192.168.2.2)
;; WHEN: Sat Nov 17 00:39:19 UTC 2018
;; MSG SIZE  rcvd: 183
```

#### CoreDNS
```
; <<>> DiG 9.11.3-1ubuntu1.2-Ubuntu <<>> service-a.aws.example.com @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33401
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 4, ADDITIONAL: 0

;; QUESTION SECTION:
;service-a.aws.example.com. IN A

;; ANSWER SECTION:
service-a.aws.example.com. 60 IN CNAME service-a-1285624372.us-west-2.elb.amazonaws.com.
service-a-1285624372.us-west-2.elb.amazonaws.com. 60 IN	A 192.168.44.44
service-a-1285624372.us-west-2.elb.amazonaws.com. 60 IN	A 192.168.55.55
service-a-1285624372.us-west-2.elb.amazonaws.com. 60 IN	A 192.168.66.66

;; AUTHORITY SECTION:
example.com.		60	IN	NS	ns-1536.awsdns-00.co.uk.
example.com.		60	IN	NS	ns-0.awsdns-00.com.
example.com.		60	IN	NS	ns-1024.awsdns-00.org.
example.com.		60	IN	NS	ns-512.awsdns-00.net.

;; Query time: 10 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sat Nov 17 00:41:57 UTC 2018
;; MSG SIZE  rcvd: 604
```

Now I've found that I can only replicate this behavior on the 3 AZ records. I have other ALBs that are only in two AZs and the resolve via `Resolv::DNS`.

```
irb(main):007:0* ::Resolv::DNS.new(nameserver: '192.168.2.2').getaddress('service-b.aws.example.com')
=> #<Resolv::IPv4 192.168.77.77>
irb(main):008:0> ::Resolv::DNS.new(nameserver: '127.0.0.1').getaddress('service-b.aws.example.com')
=> #<Resolv::IPv4 192.168.77.77>
irb(main):009:0> 
```

```
; <<>> DiG 9.11.3-1ubuntu1.2-Ubuntu <<>> service-b.aws.example.com @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 50496
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 4, ADDITIONAL: 0

;; QUESTION SECTION:
;service-b.aws.example.com.		IN	A

;; ANSWER SECTION:
service-b.aws.example.com.	60	IN	CNAME	service-b-639095310.us-west-2.elb.amazonaws.com.
service-b-639095310.us-west-2.elb.amazonaws.com. 60 IN A 192.168.77.77
service-b-639095310.us-west-2.elb.amazonaws.com. 60 IN A 192.168.88.88

;; AUTHORITY SECTION:
example.com.		60	IN	NS	ns-1536.awsdns-00.co.uk.
example.com.		60	IN	NS	ns-0.awsdns-00.com.
example.com.		60	IN	NS	ns-1024.awsdns-00.org.
example.com.		60	IN	NS	ns-512.awsdns-00.net.

;; Query time: 10 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sat Nov 17 00:49:45 UTC 2018
;; MSG SIZE  rcvd: 493
```

I'm able to see at a UDP level, CoreDNS is responding but for some reason `Resolv::DNS` isn't seeing the response on the larger responses.

```
# bg
[1]+ tcpdump -i lo host 127.0.0.1 and udp port 53 &
host:~# /opt/chef/embedded/bin/irb 
irb(main):001:0> require 'resolv'
=> true
irb(main):002:0> ::Resolv::DNS.new(nameserver: '127.0.0.1').getaddress('service-a.aws.example.com')
00:54:59.011534 IP localhost.49683 > localhost.domain: 54264+ A? service-a.aws.example.com. (57)
00:54:59.020279 IP localhost.domain > localhost.49683: 54264* 4/4/0 CNAME service-a-1285624372.us-west-2.elb.amazonaws.com., A 192.168.44.44, A 192.168.55.55, A 192.168.66.66 (604)
00:55:04.016748 IP localhost.49683 > localhost.domain: 54264+ A? service-a.aws.example.com. (57)
00:55:04.017061 IP localhost.domain > localhost.49683: 54264 4/4/0 CNAME service-a-1285624372.us-west-2.elb.amazonaws.com., A 192.168.44.44, A 192.168.55.55, A 192.168.66.66 (604)
00:55:14.026929 IP localhost.49683 > localhost.domain: 54264+ A? service-a.aws.example.com. (57)
00:55:14.027092 IP localhost.domain > localhost.49683: 54264 4/4/0 CNAME service-a-1285624372.us-west-2.elb.amazonaws.com., A 192.168.44.44, A 192.168.55.55, A 192.168.66.66 (604)
00:55:34.032603 IP localhost.49683 > localhost.domain: 54264+ A? service-a.aws.example.com. (57)
00:55:34.032786 IP localhost.domain > localhost.49683: 54264 4/4/0 CNAME service-a-1285624372.us-west-2.elb.amazonaws.com., A 192.168.44.44, A 192.168.55.55, A 192.168.66.66 (604)
Resolv::ResolvError: DNS result has no information for service-a.aws.example.com
	from /opt/chef/embedded/lib/ruby/2.4.0/resolv.rb:384:in `getaddress'
	from (irb):2
	from /opt/chef/embedded/bin/irb:11:in `<main>'
irb(main):003:0> 
```

This may be related to #2041 which cited peculiar responses with Resolv::DNS also.

**Comment by miekg** [comment]
Can you provide tcpdumps?

**Comment by miekg** [comment]
I tried replicating this with my feeble ruby knowledge, but not having
direct access to these records makes it hard. Can you open that up so I can
see those responses?

**Comment by mattlqx** [comment]
I setup an external env with the same setup and discovered that I couldn't reproduce myself on latest (1.2.6). I switched to running 1.2.4 (which is what version the CoreDNS is in the report issue) and found that the problem was reproducible. For giggles, I switched to 1.2.5 and found that worked without problem. So some change between releases 1.2.4 and 1.2.5 fixed the issue! Silly me for not just upgrading my deployments to the latest version before reporting. Sorry!
